### ISSUE TITLE
Using the copier template with conda environments

### ISSUE THREAD

**Issue by tloredo** [lincc-frameworks/python-project-template#168]
## TL;DR

In the `pipx` section of the docs, there is this remark concerning using `pipx` and `copier` with conda:

> In the [Getting Started](https://lincc-ppt.readthedocs.io/en/latest/source/overview.html) section, we assume the use of pipx. But you can use conda or pip if you’d rather have copier installed in a specific environment. We have not yet documented the differences to the getting started workflow that this would require.

Here is what conda users most likely would want to do:
```
$ conda install -c conda-forge copier
```
That's it! That installs `copier` safely isolated within the current conda env, requiring no new wrangling of the user's `.bashrc` or `.bash_profile`, and no creation of new directory trees in the user's home directory. I've successfully set up a new project using such a `copier` installation.

I feel that the *Getting started* section should say something more about `pipx`, at least forcing conda users to read the `pipx` section of the docs so they don't freak out when they see the recommended installation process putting things outside of a conda env and mucking with their shell files—just the kind of thing conda was invented to avoid.

It might be tempting for a conda user to try this instead:
```
$ conda install -c conda-forge pipx
$ pipx install copier
```
That will "work" in the sense of having essentially the same final result as installing `pipx` with `pip`. But it will not isolate the installation, and will require extra work from the user. In particular, the `pipx install copier` command results in this message:
```
installed package copier 7.1.0, installed using Python 3.10.10
  These apps are now globally available
    - copier
⚠️  Note: '/Users/loredo/.local/bin' is not on your PATH environment variable.
    These apps will not be globally accessible until your PATH is updated. Run
    `pipx ensurepath` to automatically add it, or manually modify your PATH in
    your shell's config file (i.e. ~/.bashrc).
done!
```
So not only will content be installed outside the environment, but also the user will have to manually update their `PATH` definition. So conda users probably will want to just install `copier` directly from conda-forge. As far as I can tell, there is no need for `pipx` at all.

## Background

As I believe is a common practice, I use miniconda and I do all of my work in conda environments, with different projects using different Python versions (including some old ones still using Python 2). I had never before encountered `pipx`.

I followed the instructions in the *Getting started* section, working within a conda env setup for developing a new project. Since I didn't have `pipx` in the env, when I came to this,

> pipx (recommended - more info about pipx [here](https://lincc-ppt.readthedocs.io/en/latest/practices/pipx.html))

instead of going to the link, I went right to the `pipx` docs to install it using its own instructions.

I slightly freaked out when I saw it mucked with `PATH` and that a subsequent `copier` install created a tree outside of my conda env and put `copier` there. This mode of operation is not explained clearly in the `pipx` documentation (I later found it explained much more clearly in the LINCC template docs, but that was too late to be useful). This seems to me to work against the spirit (and wisdom) of conda environments. My various envs install very different Python environments; it seems unwise to install Python content outside an env. And in any case, neither my project nor the LINCC template needs such a global installation. I don't see a need for the recommended setup to do anything beyond the least invasive thing possible.

At the very least, I feel the docs should change that `pipx` line to read:

> pipx (recommended - more info about pipx [here](https://lincc-ppt.readthedocs.io/en/latest/practices/pipx.html), conda environment users should be sure to read that section)

The `pipx` section of the LINCC template docs are really helpful—they explain what `pipx` is for better than the `pipx` docs do. Maybe you should suggest revisions to the `pipx` docs based on your content.

**Comment by gschwend** [comment]
thanks for sharing this @tloredo!

I am using conda on a linux pc and the command 
```$ conda install -c conda-forge copier``` 
simply does not work. It takes forever at "Solving environment" step. 

Following LINCC template docs suggestion, I went to pipx documentation, and there I found the commands for installation via pip: 
``` 
python3 -m pip install --user pipx
python3 -m pipx ensurepath
``` 
I run them both inside my particular conda environment and, as you pointed above, the second line made pipx to appear in `.local/bin` regardless of my conda env is active or not. 
Is that a bad thing? Should I redo this in another way?

**Comment by drewoldag** [comment]
Pipx operates a little differently than your typical tool. It is considered optional, as there are other ways of acquiring `copier`. We advocated for `pipx` because we believed it was easier for a new-to-python user to get started with. 

From a high level, `pipx` is meant to install specific new packages, like `copier` in an isolated, but globally available, site-package (along with any dependencies necessary). The purpose, in this specific example, would be to allow a user to call `copier` from any environment. The additional benefit of using `pipx` is that is makes it easy to cleanly uninstall any package that was installed using it. For instance, if you call `pipx install copier`. Then `copier` and it's dependencies will be placed in an isolated, globally available location. Then calling `pipx uninstall copier` will completely remove `copier` and it's dependencies without unintentionally affecting any other packages.

If there are obvious ways that we can improve the documentation or workflow to make it more clear, please let us know, the feedback is very much appreciated!

**Comment by drewoldag** [comment]
And @gschwend to answer your last two questions, I believe that what you have configured is correct. So there shouldn't be any need for rework.

**Comment by tloredo** [comment]
@gschwend Hi Julia! Two points:

First: The installation outside of the environment is intentional behavior for `pipx` and `pipx`-installed packages. The idea is to keep widely-used command-line tools always available, in separate environments with their own dependencies (`pipx` and any `pipx`-installed packages get their own environment, and work happily adjacent to any other environment you may be working in). This lets you use those relatively simple tools with a single installation parallel to any other environments you may be using (via `conda` or whatever). If I understand things correctly, the potential for such tools doing something incompatible with another environment is quite unlikely. That slightly freaked me out at first, but Drew helped me understand how `pipx` works during LINCC office hours, and if I were doing my installation over again, I *might* just do the recommended `pipx` installation. Though to be honest, I still prefer keeping everything isolated, including command-line tools.

Second: The "solving environment" issue you are experiencing is probably something worth addressing, even if you are content sticking with `pipx` for what you're currently working on. I don't have a very deep understanding of what causes such problems with `conda`, but here are some things I discovered through web searches and experimenting.

* I had an Anaconda installation that I'd been maintaining for several years, with many of my environments themselves also using the `anaconda` set of packages. The solution times kept growing. About a year ago I hosed the whole installation, started with *Miniconda* as my base environment, and for most of my environments I now use a fairly minimal set of packages. I maintain a couple full Anaconda environments for quick experimenting, but for concrete projects, I use smaller, focused environments. Astropy has a recommended minimal environment that often start with (I add `ipython` and `jupyter` to it). Simpler environments make dependency solving faster.
* There's a `mamba` project that arose a few years ago to offer a faster environment solver than `conda`. I haven't used it, and I believe a lot of its innovations are now included in `conda` (make sure you often do `conda update conda` to get such updates), so there may not be a lot of value in trying it. But it's worth knowing about.
* Finally, using the right `conda` channels in the right search order is *very important*. I believe the current recommendation is to switch from the default channel being first to instead prioritizing `conda-forge`. Since I've done that, I've never had problems with long environment solutions. This is what's in my `.condarc` file to implement this:

```
channel_priority: flexible
channels:
  - conda-forge
  - defaults
```

Changing to that probably won't fix problems with existing environments (which may have a bad mix of packages from different channels), but it may make future envs easier to maintain.

I hope some of this will be helpful!

PS: Hosing my conda installation wasn't painful because I've been pretty meticulous about keeping a log of my environment definitions (I do it in a Joplin notebook, synced across my machines). So I was able to easily re-create any old environments I needed once I started over. It would probably be better to use `environment.yml` files or something like that, but for now it's been easy enough to just keep a log of my `conda install ...` commands. For an environment that may be shared with others, using an environment definition file would be better!

**Comment by gschwend** [comment]
Wow! That's a lot of valuable tips. Thank you so much!

**Comment by drewoldag** [comment]
I've come back to this issue after quite a long break, and after re-reading the thread it seems like the main request here is to add a little bit more to the documentation around pipx in the recommendations section.
